### PR TITLE
Add JSONification of Soroban invocation trees.

### PR DESCRIPTION
**Pull request by Shaptic**

### CODE REVIEW

File: src/invocation.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
+import { Address } from './address';
+import { scValToNative } from './scval';
+
+/**
+ * @typedef CreateInvocation
```
Comment (paulbellamy) [inline src/invocation.js:5]: are these actual types somewhere?

File: src/invocation.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+/**
+ * Turns a raw invocation tree into a human-readable format.
+ *
+ * This is designed to make the invocation tree easier to understand in order to
+ * inform users about the side-effects of their contract calls. This will help
+ * make informed decisions about whether or not a particular invocation will
+ * result in what you expect it to.
+ *
+ * @param {xdr.SorobanAuthorizedInvocation} root  the raw XDR of the invocation,
```
Comment (paulbellamy) [inline src/invocation.js:35]: What do you mean by "raw XDR"? base64 string? js object?

File: src/invocation.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+ *    likely acquired from transaction simulation. this is either from the
```
Comment (paulbellamy) [inline src/invocation.js:36]: Maybe an example of how you would acquire it? or parse it? or something?

File: src/invocation.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+ *    {@link Operation.invokeHostFunction} itself (the `func` field), or from
+ *    the authorization entries ({@link SorobanAuthorizationEntry}, the
+ *    `rootInvocation` field)
+ *
+ * @returns {InvocationTree}  a human-readable version of the invocation tree
+ */
+export function buildInvocationTree(root) {
+  /** @type {InvocationTree} */
```
Comment (paulbellamy) [inline src/invocation.js:44]: Again, I'm assuming this type should be defined in `types`?

**Review by paulbellamy — COMMENTED** [review]
I'm not really clear what the point of this is. Maybe a before/after example would help?

The JSON produces is *marginally* more human-friendly, but I'd imagine wallets/apps would be walking through the SorobanAuthorizedInvocation to produce HTML out the other end. This fn doesn't seem super useful for that. Maybe a `walkInvocationTree(fn)` would be more useful for that.

Generally, I've found it more useful to try to do stuff in the dapp example, and then see what helpers I can extract from there into the sdk/base. That way you have a motivating example and can make sure you've got an abstraction that is at least useful.

But idk tho.

**Comment by Shaptic** [comment]
Thank you for the review, @paulbellamy :pray:

> try to do stuff in the dapp example [...] motivating example [...]

I forgot to include the context that motivated this: [Slack thread](https://stellarfoundation.slack.com/archives/C016CPSLVLG/p1691616428989009). The idea is for Freighter et al. to have meaningful popups when asking users to authorize invocation trees.

> The JSON produces is _marginally_ more human-friendly

**Hard** disagree, but this isn't obvious without an example. Flattening the structure and leveraging `scValToNative` makes a **huge** difference in readability. Which one is easier to grok? (both are `JSON.stringify`d)

#### XDR
<details>

```xdr
{
  "_attributes": {
    "function": {
      "_switch": {
        "name": "sorobanAuthorizedFunctionTypeContractFn",
        "value": 0
      },
      "_arm": "contractFn",
      "_value": {
        "_attributes": {
          "contractAddress": {
            "_switch": {
              "name": "scAddressTypeContract",
              "value": 1
            },
            "_arm": "contractId",
            "_armType": {
              "_length": 32
            },
            "_value": {
              "type": "Buffer",
              "data": [
                232,
                8,
                251,
                255,
                185,
                70,
                175,
                127,
                248,
                65,
                1,
                210,
                82,
                54,
                174,
                141,
                74,
                125,
                242,
                129,
                116,
                44,
                34,
                107,
                197,
                3,
                127,
                111,
                110,
                162,
                113,
                83
              ]
            }
          },
          "functionName": "purchase",
          "args": [
            {
              "_switch": {
                "name": "scvString",
                "value": 14
              },
              "_arm": "str",
              "_armType": {
                "_maxLength": 4294967295
              },
              "_value": "SomeNft:GBS7D7AJFWKIMWLD7KFAV2AXHSKCOE4DSSVKZQBGJCC3FRGS7WAS5WA6"
            },
            {
              "_switch": {
                "name": "scvU64",
                "value": 5
              },
              "_arm": "u64",
              "_value": {
                "_value": "7"
              }
            }
          ]
        }
      }
    },
    "subInvocations": [
      {
        "_attributes": {
          "function": {
            "_switch": {
              "name": "sorobanAuthorizedFunctionTypeContractFn",
              "value": 0
            },
            "_arm": "contractFn",
            "_value": {
              "_attributes": {
                "contractAddress": {
                  "_switch": {
                    "name": "scAddressTypeContract",
                    "value": 1
                  },
                  "_arm": "contractId",
                  "_armType": {
                    "_length": 32
                  },
                  "_value": {
                    "type": "Buffer",
                    "data": [
                      83,
                      50,
                      135,
                      62,
                      59,
                      72,
                      227,
                      3,
                      235,
                      42,
                      185,
                      176,
                      229,
                      104,
                      99,
                      81,
                      154,
                      204,
                      198,
                      232,
                      58,
                      194,
                      20,
                      154,
                      167,
                      60,
                      88,
                      140,
                      145,
                      89,
                      180,
                      48
                    ]
                  }
                },
                "functionName": "swap",
                "args": [
                  {
                    "_switch": {
                      "name": "scvString",
                      "value": 14
                    },
                    "_arm": "str",
                    "_armType": {
                      "_maxLength": 4294967295
                    },
                    "_value": "native"
                  },
                  {
                    "_switch": {
                      "name": "scvString",
                      "value": 14
                    },
                    "_arm": "str",
                    "_armType": {
                      "_maxLength": 4294967295
                    },
                    "_value": "USDC:GDEWIEQEB657FDO3GZIUTYS57T4M65QP352A2PN4CJPBCR55I5FWVAKG"
                  },
                  {
                    "_switch": {
                      "name": "scvAddress",
                      "value": 18
                    },
                    "_arm": "address",
                    "_value": {
                      "_switch": {
                        "name": "scAddressTypeAccount",
                        "value": 0
                      },
                      "_arm": "accountId",
                      "_value": {
                        "_switch": {
                          "name": "publicKeyTypeEd25519",
                          "value": 0
                        },
                        "_arm": "ed25519",
                        "_armType": {
                          "_length": 32
                        },
                        "_value": {
                          "type": "Buffer",
                          "data": [
                            76,
                            219,
                            245,
                            104,
                            41,
                            198,
                            51,
                            53,
                            239,
                            66,
                            121,
                            235,
                            61,
                            236,
                            110,
                            210,
                            177,
                            117,
                            102,
                            181,
                            16,
                            35,
                            129,
                            214,
                            69,
                            128,
                            154,
                            252,
                            64,
                            79,
                            52,
                            127
                          ]
                        }
                      }
                    }
                  },
                  {
                    "_switch": {
                      "name": "scvAddress",
                      "value": 18
                    },
                    "_arm": "address",
                    "_value": {
                      "_switch": {
                        "name": "scAddressTypeAccount",
                        "value": 0
                      },
                      "_arm": "accountId",
                      "_value": {
                        "_switch": {
                          "name": "publicKeyTypeEd25519",
                          "value": 0
                        },
                        "_arm": "ed25519",
                        "_armType": {
                          "_length": 32
                        },
                        "_value": {
                          "type": "Buffer",
                          "data": [
                            52,
                            74,
                            39,
                            21,
                            56,
                            39,
                            157,
                            177,
                            87,
                            193,
                            83,
                            66,
                            151,
                            143,
                            125,
                            19,
                            6,
                            239,
                            212,
                            82,
                            148,
                            51,
                            22,
                            140,
                            81,
                            169,
                            48,
                            184,
                            248,
                            113,
                            124,
                            83
                          ]
                        }
                      }
                    }
                  }
                ]
              }
            }
          },
          "subInvocations": [
            {
              "_attributes": {
                "function": {
                  "_switch": {
                    "name": "sorobanAuthorizedFunctionTypeContractFn",
                    "value": 0
                  },
                  "_arm": "contractFn",
                  "_value": {
                    "_attributes": {
                      "contractAddress": {
                        "_switch": {
                          "name": "scAddressTypeContract",
                          "value": 1
                        },
                        "_arm": "contractId",
                        "_armType": {
                          "_length": 32
                        },
                        "_value": {
                          "type": "Buffer",
                          "data": [
                            36,
                            67,
                            12,
                            76,
                            242,
                            197,
                            130,
                            66,
                            144,
                            142,
                            29,
                            145,
                            208,
                            242,
                            207,
                            18,
                            181,
                            117,
                            253,
                            41,
                            48,
                            141,
                            3,
                            69,
                            36,
                            114,
                            219,
                            194,
                            117,
                            150,
                            132,
                            109
                          ]
                        }
                      },
                      "functionName": "transfer",
                      "args": [
                        {
                          "_switch": {
                            "name": "scvAddress",
                            "value": 18
                          },
                          "_arm": "address",
                          "_value": {
                            "_switch": {
                              "name": "scAddressTypeAccount",
                              "value": 0
                            },
                            "_arm": "accountId",
                            "_value": {
                              "_switch": {
                                "name": "publicKeyTypeEd25519",
                                "value": 0
                              },
                              "_arm": "ed25519",
                              "_armType": {
                                "_length": 32
                              },
                              "_value": {
                                "type": "Buffer",
                                "data": [
                                  76,
                                  219,
                                  245,
                                  104,
                                  41,
                                  198,
                                  51,
                                  53,
                                  239,
                                  66,
                                  121,
                                  235,
                                  61,
                                  236,
                                  110,
                                  210,
                                  177,
                                  117,
                                  102,
                                  181,
                                  16,
                                  35,
                                  129,
                                  214,
                                  69,
                                  128,
                                  154,
                                  252,
                                  64,
                                  79,
                                  52,
                                  127
                                ]
                              }
                            }
                          }
                        },
                        {
                          "_switch": {
                            "name": "scvString",
                            "value": 14
                          },
                          "_arm": "str",
                          "_armType": {
                            "_maxLength": 4294967295
                          },
                          "_value": "7"
                        }
                      ]
                    }
                  }
                },
                "subInvocations": []
              }
            },
            {
              "_attributes": {
                "function": {
                  "_switch": {
                    "name": "sorobanAuthorizedFunctionTypeContractFn",
                    "value": 0
                  },
                  "_arm": "contractFn",
                  "_value": {
                    "_attributes": {
                      "contractAddress": {
                        "_switch": {
                          "name": "scAddressTypeContract",
                          "value": 1
                        },
                        "_arm": "contractId",
                        "_armType": {
                          "_length": 32
                        },
                        "_value": {
                          "type": "Buffer",
                          "data": [
                            56,
                            133,
                            65,
                            88,
                            225,
                            40,
                            234,
                            250,
                            236,
                            145,
                            45,
                            73,
                            128,
                            111,
                            87,
                            142,
                            198,
                            124,
                            243,
                            151,
                            85,
                            5,
                            37,
                            42,
                            194,
                            53,
                            2,
                            51,
                            250,
                            227,
                            108,
                            176
                          ]
                        }
                      },
                      "functionName": "transfer",
                      "args": [
                        {
                          "_switch": {
                            "name": "scvAddress",
                            "value": 18
                          },
                          "_arm": "address",
                          "_value": {
                            "_switch": {
                              "name": "scAddressTypeAccount",
                              "value": 0
                            },
                            "_arm": "accountId",
                            "_value": {
                              "_switch": {
                                "name": "publicKeyTypeEd25519",
                                "value": 0
                              },
                              "_arm": "ed25519",
                              "_armType": {
                                "_length": 32
                              },
                              "_value": {
                                "type": "Buffer",
                                "data": [
                                  76,
                                  219,
                                  245,
                                  104,
                                  41,
                                  198,
                                  51,
                                  53,
                                  239,
                                  66,
                                  121,
                                  235,
                                  61,
                                  236,
                                  110,
                                  210,
                                  177,
                                  117,
                                  102,
                                  181,
                                  16,
                                  35,
                                  129,
                                  214,
                                  69,
                                  128,
                                  154,
                                  252,
                                  64,
                                  79,
                                  52,
                                  127
                                ]
                              }
                            }
                          }
                        },
                        {
                          "_switch": {
                            "name": "scvString",
                            "value": 14
                          },
                          "_arm": "str",
                          "_armType": {
                            "_maxLength": 4294967295
                          },
                          "_value": "7"
                        }
                      ]
                    }
                  }
                },
                "subInvocations": []
              }
            }
          ]
        }
      },
      {
        "_attributes": {
          "function": {
            "_switch": {
              "name": "sorobanAuthorizedFunctionTypeContractFn",
              "value": 0
            },
            "_arm": "contractFn",
            "_value": {
              "_attributes": {
                "contractAddress": {
                  "_switch": {
                    "name": "scAddressTypeContract",
                    "value": 1
                  },
                  "_arm": "contractId",
                  "_armType": {
                    "_length": 32
                  },
                  "_value": {
                    "type": "Buffer",
                    "data": [
                      232,
                      8,
                      251,
                      255,
                      185,
                      70,
                      175,
                      127,
                      248,
                      65,
                      1,
                      210,
                      82,
                      54,
                      174,
                      141,
                      74,
                      125,
                      242,
                      129,
                      116,
                      44,
                      34,
                      107,
                      197,
                      3,
                      127,
                      111,
                      110,
                      162,
                      113,
                      83
                    ]
                  }
                },
                "functionName": "GBGNX5LIFHDDGNPPIJ46WPPMN3JLC5LGWUICHAOWIWAJV7CAJ42H6FBG",
                "args": [
                  {
                    "_switch": {
                      "name": "scvAddress",
                      "value": 18
                    },
                    "_arm": "address",
                    "_value": {
                      "_switch": {
                        "name": "scAddressTypeAccount",
                        "value": 0
                      },
                      "_arm": "accountId",
                      "_value": {
                        "_switch": {
                          "name": "publicKeyTypeEd25519",
                          "value": 0
                        },
                        "_arm": "ed25519",
                        "_armType": {
                          "_length": 32
                        },
                        "_value": {
                          "type": "Buffer",
                          "data": [
                            76,
                            219,
                            245,
                            104,
                            41,
                            198,
                            51,
                            53,
                            239,
                            66,
                            121,
                            235,
                            61,
                            236,
                            110,
                            210,
                            177,
                            117,
                            102,
                            181,
                            16,
                            35,
                            129,
                            214,
                            69,
                            128,
                            154,
                            252,
                            64,
                            79,
                            52,
                            127
                          ]
                        }
                      }
                    }
                  },
                  {
                    "_switch": {
                      "name": "scvU64",
                      "value": 5
                    },
                    "_arm": "u64",
                    "_value": {
                      "_value": "1"
                    }
                  }
                ]
              }
            }
          },
          "subInvocations": []
        }
      }
    ]
  }
}
</details>

#### JSON

<details>

```json
{
  "type": "execute",
  "args": {
    "source": "CDUAR677XFDK677YIEA5EURWV2GUU7PSQF2CYITLYUBX633OUJYVGTSS",
    "function": "purchase",
    "args": [
      "SomeNft:GBS7D7AJFWKIMWLD7KFAV2AXHSKCOE4DSSVKZQBGJCC3FRGS7WAS5WA6",
      "7"
    ]
  },
  "subInvocations": [
    {
      "type": "execute",
      "args": {
        "source": "CBJTFBZ6HNEOGA7LFK43BZLIMNIZVTGG5A5MEFE2U46FRDERLG2DAJTB",
        "function": "swap",
        "args": [
          "native",
          "USDC:GDEWIEQEB657FDO3GZIUTYS57T4M65QP352A2PN4CJPBCR55I5FWVAKG",
          "GBGNX5LIFHDDGNPPIJ46WPPMN3JLC5LGWUICHAOWIWAJV7CAJ42H6FBG",
          "GA2EUJYVHATZ3MKXYFJUFF4PPUJQN36UKKKDGFUMKGUTBOHYOF6FHSUZ"
        ]
      },
      "subInvocations": [
        {
          "type": "execute",
          "args": {
            "source": "CASEGDCM6LCYEQUQRYOZDUHSZ4JLK5P5FEYI2A2FERZNXQTVS2CG3BAH",
            "function": "transfer",
            "args": [
              "GBGNX5LIFHDDGNPPIJ46WPPMN3JLC5LGWUICHAOWIWAJV7CAJ42H6FBG",
              "7"
            ]
          },
          "subInvocations": []
        },
        {
          "type": "execute",
          "args": {
            "source": "CA4IKQKY4EUOV6XMSEWUTADPK6HMM7HTS5KQKJJKYI2QEM724NWLAKEI",
            "function": "transfer",
            "args": [
              "GBGNX5LIFHDDGNPPIJ46WPPMN3JLC5LGWUICHAOWIWAJV7CAJ42H6FBG",
              "7"
            ]
          },
          "subInvocations": []
        }
      ]
    },
    {
      "type": "execute",
      "args": {
        "source": "CDUAR677XFDK677YIEA5EURWV2GUU7PSQF2CYITLYUBX633OUJYVGTSS",
        "function": "GBGNX5LIFHDDGNPPIJ46WPPMN3JLC5LGWUICHAOWIWAJV7CAJ42H6FBG",
        "args": [
          "GBGNX5LIFHDDGNPPIJ46WPPMN3JLC5LGWUICHAOWIWAJV7CAJ42H6FBG",
          "1"
        ]
      },
      "subInvocations": []
    }
  ]
}
```

</details>

File: src/invocation.js (hunk 1 above)
Comment (Shaptic) [reply src/invocation.js:5]: this is just an api proposal, so I was trying to use jsdocs to define the types

File: src/invocation.js (hunk 2 above)
Comment (Shaptic) [reply src/invocation.js:35]: the js structure, yeah. string could work too but this is just outlining the api more than a full solution

**Comment by paulbellamy** [comment]
Putting on my, "non technical user" hat. Both the xdr and json examples you gave look like unintelligible computer-nonsense. Fine, the json outputs contract IDs instead of byte arrays, etc. 🤷 

Interestingly, when you actually wanted to explain what the tree structure you were building does you invented a different (more graphical) format to do that. 😉 

```
 purchase("SomeNft:G...", 7 xlm)
     |
     +--- swap(xlm, usdc, from, to)
     |      |
     |      |- xlm.transfer(from, to, 7)
     |      |
     |      |- usdc.transfer(to, from, 1)
     |
     +--- someNft.transfer(invoker, contract, 1)
```

^ This is what I would expect wallets to do. So if `xdr -> json -> html` is easier for wallets/dapps than `xdr -> html`, then fine, I guess.

**Comment by paulbellamy** [comment]
To clarify, by "xdr" in the above comment, I mean the js classes from `stellar-base`, not like base64 strings. But sounds like an intermediate representation might still be helpful. 👍

File: src/invocation.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+  /** @type {xdr.CreateContractArgs | xdr.InvokeContractArgs} */
+  const inner = fn.value();
+
+  switch (fn.switch().value) {
+    // sorobanAuthorizedFunctionTypeContractFn
+    case 0:
+      output.type = 'execute';
+      output.args = {
+        source: new Address.fromScAddress(inner.contractAddress()).toString(),
```
Comment (aristidesstaffieri) [inline src/invocation.js:58]: no new needed here

File: src/invocation.js [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        function: inner.functionName(),
+        args: inner.args().map((arg) => scValToNative(arg))
```
Comment (aristidesstaffieri) [inline src/invocation.js:60]: assuming you might take this and render the json directly in the dom, does it make sense to stringify the args like in the case of `BigInt`.

` args: inner.args().map((arg) => scValToNative(arg).toString())`
this worked well for me when trying this in Freighter on an AuthEntry invocation.

Comment (Shaptic) [reply src/invocation.js:60]: That's interesting :thinking: 

At this layer, I wouldn't want to do more interpretation than is necessary to get the point across (e.g. coalescing the int types to BigInt is good, but moving to string creates ambiguity with actual stringified values, maybe? or maybe I'm overthinking this). 

At a higher layer, you can already accomplish this by walking the returned object:

```javascript
const friendly = buildInvocationTree(someAuthEntry.rootInvocation());

alert(
  JSON.stringify(
    friendly, 
    // stringify BigInts
    (_, val) => typeof val === 'bigint' ? val.toString() : val, 
    // 2-space indentation
    2)
);
```

**Review by aristidesstaffieri — APPROVED** [review]

**Pull request merged**